lib/ack-tracker/batched_ack_tracker.c

```diff
@@ -0,0 +1,300 @@
+  GMutex acked_records_lock;
+  gulong acked_records_num;
+  GList *acked_records;
+  struct iv_timer batch_timer;
+  struct iv_event request_restart_timer;
+  gboolean watches_running;
+} BatchedAckTracker;
+
+static void
+_ack_record_free(AckRecord *s)
+{
+  BatchedAckRecord *self = (BatchedAckRecord *) s;
+  bookmark_destroy(&self->super.bookmark);
+
+  g_free(self);
+}
+
+static void
+_ack_batch(BatchedAckTracker *self, GList *ack_records)
+{
+  self->on_batch_acked.func(ack_records, self->on_batch_acked.user_data);
+  g_list_free_full(ack_records, (GDestroyNotify) _ack_record_free);
+}
+
+static Bookmark *
+_request_bookmark(AckTracker *s)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) s;
+  if (!self->pending_ack_record)
+    self->pending_ack_record = g_new0(BatchedAckRecord, 1);
+
+  return &(self->pending_ack_record->super.bookmark);
+}
+
+static void
+_assign_pending_ack_record_to_msg(BatchedAckTracker *self, LogMessage *msg)
+{
+  LogSource *source = self->super.source;
+  log_pipe_ref((LogPipe *) source);
+  self->pending_ack_record->super.bookmark.persist_state = source->super.cfg->state;
+  self->pending_ack_record->super.tracker = &self->super;
+  msg->ack_record = (AckRecord *) self->pending_ack_record;
+}
+
+static void
+_track_msg(AckTracker *s, LogMessage *msg)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *)s;
+  _assign_pending_ack_record_to_msg(self, msg);
+  self->pending_ack_record = NULL;
+}
+
+static GList *
+_append_ack_record_to_batch(BatchedAckTracker *self, AckRecord *ack_record)
+{
+  GList *full_batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    self->acked_records = g_list_prepend(self->acked_records, ack_record);
+    ++self->acked_records_num;
+    if (self->acked_records_num == self->batch_size)
+      {
+        full_batch = self->acked_records;
+        self->acked_records = NULL;
+        self->acked_records_num = 0;
+      }
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+
+  return full_batch;
+}
+
+static void
+_stop_batch_timer(BatchedAckTracker *self)
+{
+  if (iv_timer_registered(&self->batch_timer))
+    iv_timer_unregister(&self->batch_timer);
+}
+
+static void
+_start_batch_timer(BatchedAckTracker *self)
+{
+  if (self->timeout <= 0)
+    return;
+
+  iv_validate_now();
+  self->batch_timer.expires = iv_now;
+  timespec_add_msec(&self->batch_timer.expires, self->timeout);
+
+  iv_timer_register(&self->batch_timer);
+}
+
+static void
+_restart_batch_timer(BatchedAckTracker *self)
+{
+  _stop_batch_timer(self);
+  _start_batch_timer(self);
+}
+
+static void
+_batch_timeout(gpointer data)
+{
+  BatchedAckTracker *self = (BatchedAckTracker *) data;
+  GList *batch = NULL;
+  g_mutex_lock(&self->acked_records_lock);
+  {
+    batch = self->acked_records;
+    self->acked_records = NULL;
+    self->acked_records_num = 0;
+  }
+  g_mutex_unlock(&self->acked_records_lock);
+  if (batch)
+    {
+      _ack_batch(self, batch);
+      _start_batch_timer(self);
```

Comment:
Sorry, I've found another one:

I think the timer should be restarted even when `batch` is empty.
In case there are no messages for `batch_timeout` millisecond, the timer won't be restarted, so we'll wait for a full batch after the inactive time period.

Reply:
you are right, I was added that line at the wrong place - it should be re-registered unconditionally

I fixed it not, please check it.